include/sddf/block/blk_shared_queue.h

```diff
@@ -0,0 +1,342 @@
+}
+
+/**
+ * Check if the response queue is empty.
+ *
+ * @param queue_handle queue handle containing response queue.
+ *
+ * @return true indicates the response queue is empty, false otherwise.
+ */
+static inline bool blk_resp_queue_empty(blk_queue_handle_t *queue_handle)
+{
+    return !((queue_handle->resp_queue->write_idx - queue_handle->resp_queue->read_idx) % queue_handle->resp_queue->size);
+}
+
+/**
+ * Check if the request queue is full.
+ *
+ * @param queue_handle queue handle containing request queue.
+ *
+ * @return true indicates the request queue is full, false otherwise.
+ */
+static inline bool blk_req_queue_full(blk_queue_handle_t *queue_handle)
+{
+    return !((queue_handle->req_queue->write_idx - queue_handle->req_queue->read_idx + 1) % queue_handle->req_queue->size);
+}
+
+/**
+ * Check if the response queue is full.
+ *
+ * @param queue_handle queue handle containing response queue.
+ *
+ * @return true indicates the response queue is full, false otherwise.
+ */
+static inline bool blk_resp_queue_full(blk_queue_handle_t *queue_handle)
+{
+    return !((queue_handle->resp_queue->write_idx - queue_handle->resp_queue->read_idx + 1) % queue_handle->resp_queue->size);
+}
+
+/**
+ * Get the number of elements in a request queue.
+ *
+ * @param queue_handle queue handle containing request and response queues.
+ *
+ * @return number of elements in the queue.
+ */
+static inline int blk_req_queue_size(blk_queue_handle_t *queue_handle)
+{
+    return (queue_handle->req_queue->write_idx - queue_handle->req_queue->read_idx);
+}
+
+/**
+ * Get the number of elements in a response queue.
+ *
+ * @param queue_handle queue handle containing request and response queues.
+ *
+ * @return number of elements in the queue.
+ */
+static inline int blk_resp_queue_size(blk_queue_handle_t *queue_handle)
+{
+    return (queue_handle->resp_queue->write_idx - queue_handle->resp_queue->read_idx);
+}
+
+/**
+ * Enqueue an element into the request queue.
+ *
+ * @param queue_handle queue handle containing request queue to enqueue to.
+ * @param code request code.
+ * @param addr encoded dma address of data to read/write.
+ * @param block_number block number to read/write to.
+ * @param count the number of blocks to read/write
+ * @param id request ID to identify this request.
+ *
+ * @return -1 when request queue is full, 0 on success.
+ */
+static inline int blk_enqueue_req(blk_queue_handle_t *queue_handle,
+                                        blk_request_code_t code,
+                                        uintptr_t addr,
+                                        uint32_t block_number,
+                                        uint16_t count,
+                                        uint32_t id)
+{
+    if (blk_req_queue_full(queue_handle)) {
+        return -1;
+    }
+
+    queue_handle->req_queue->buffers[queue_handle->req_queue->write_idx % queue_handle->req_queue->size].code = code;
```

Comment:
Might be worth it to have a macro for `queue_handle->req_queue->buffers[queue_handle->req_queue->write_idx % queue_handle->req_queue->size]`